i18n: Create new function `addLocaleData` to merge domain configuration

arthur791004 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

As https://github.com/WordPress/gutenberg/issues/34101 and https://github.com/WordPress/gutenberg/pull/37686 mentioned, calling `setLocaleData` multiple times, the domain configuration will be overwritten by the latest one. It might be better to merge the domain configuration rather than overwrite. However, we should avoid any breaking change. So, create a new function `addLocaleData` to merge the domain configuration and keep the behavior of `setLocaleData` the same.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Go to http://localhost:8888/wp-admin/post.php?post=1&action=edit
2. Open Developer Tool
3. Do the following commands
   ```js
   wp.i18n.addLocaleData({ "": { lang: 'en' } })
   wp.i18n.addLocaleData({ "": {} })
   wp.i18n.getLocaleData()[""]
   ```
4. Output
   ```js
   { lang: "en", plural_forms: f } 
   ```

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

arthur791004 committed:
i18n: Add addLocaleData function

arthur791004 committed:
i18n: Add test case for addLocaleData

sirreal commented:
👋  Thanks for pinging me, @swissspidy.

I haven't been working with this package lately. I'm not in a good position to make decisions about introducing new APIs.

Once we're confident we'd like to introduce a new API, I'm happy to help with code review.

jsnajdr commented on packages/i18n/src/create-i18n.js:
```diff
@@ -186,12 +200,37 @@ export const createI18n = ( initialData, initialDomain, hooks ) => {
+	 */
+	const doAddLocaleData = ( data, domain = 'default' ) => {
+		tannin.data[ domain ] = {
+			...DEFAULT_LOCALE_DATA,
```
The `...DEFAULT_LOCALE_DATA` spread is redundant, because it has only the `''` key and that's gurananteed to be completely overwritten by the `''` key few lines later. The `doSetLocaleData` function has the same redundancy.

jsnajdr commented on packages/i18n/src/create-i18n.js:
```diff
@@ -186,12 +200,37 @@ export const createI18n = ( initialData, initialDomain, hooks ) => {
+			'': {
+				...DEFAULT_LOCALE_DATA[ '' ],
+				...( tannin.data[ domain ] || {} )[ '' ],
+				...( data || {} )[ '' ],
```
You can use optional chaining for this: `...tannin.data[ domain ]?.[ '' ]`

jsnajdr commented on packages/i18n/src/create-i18n.js:
```diff
@@ -186,12 +200,37 @@ export const createI18n = ( initialData, initialDomain, hooks ) => {
 
+	/** @type {AddLocaleData} */
+	const addLocaleData = ( data, domain ) => {
+		doAddLocaleData( data, domain );
```
The `doAddLocaleData` function could be inlined here. There is a standalone `doSetLocaleData` function because we sometimes want to do the operation without firing notifications, but `addLocaleData` doesn't have that requirement.

jsnajdr replied:
Now you can remove `doAddLocaleData` completely, and do the `tannin` assignment directly in the `addLocaleData` body.

jsnajdr reviewed (approved):
This is a great addition, thank you! The new API makes sense.

There's a pre-existing little bug in both `setLocaleData` and `addLocaleData`: as they also support updating the locale "metadata" under the `''` key, they can also update the `[ '' ][ 'plural-forms' ]` key. But the `tannin` library internally caches that. Both functions should therefore do:
```js
tannin.pluralForms[ domain ] = null;
```
to force updating the cache. That would be nice to fix, either here or in a follow-up PR.

arthur791004 committed:
Remove redundant spread

arthur791004 committed:
Change to use optional chaining

arthur791004 committed:
Make doAddLocaleData an inline function

arthur791004 committed:
Clean up pluralForms cache

arthur791004 commented:
@jsnajdr Thanks for your review! I've fixed `pluralForms` cache issue and made some changes according to your feedback 👍

As the value of `tannin.pluralForms[ domain ]` should be `function`, I deleted the value to force updating the cache.

jsnajdr reviewed (approved):
Thanks for addressing the feedback, the PR looks good :+1: Do you need any help with merging or do you have permission to do that yourself.

There are e2e failures in `editor/blocks/navigation.test.js` -- are they possibly related? The test fails to find an element by it text content, where the text is produced by `__()`, so `__()` malfunction _is_ a possible reason for failure.

arthur791004 committed:
Remove doAddLocaleData function

arthur791004 commented:
> Do you need any help with merging or do you have permission to do that yourself.

@jsnajdr  Oh yes, I need help to merge this one as I don't have permission. Thanks!

> There are e2e failures in editor/blocks/navigation.test.js -- are they possibly related?

I don't think the e2e failures are related to this one as everything keeps the same except removing redundant 
 spread and fixing the cache issue. But I think both of them won't affect the e2e testing 🤔

swissspidy commented:
A changelog entry for this new function would be great

arthur791004 committed:
Update changelog files

arthur791004 commented:
> A changelog entry for this new function would be great

@swissspidy I've updated the changelog, thanks 👍

jsnajdr commented on packages/i18n/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 ## Unreleased
 
+- Add new `addLocaleData` method to merges locale data into the Tannin instance by domain. Note that this function will also merges the domain configuration.
+
```
Let's fix the typo in "merges" which should read as "merge". And I would personally remove the "Note" because a changelog entry doesn't need this level of detail about the new method.

The PR will be finally ready to land after this last fixup is done.

arthur791004 replied:
Oops...I'm sorry for the typo 😅. 

> And I would personally remove the "Note" because a changelog entry doesn't need this level of detail about the new method.

Okay 👍

arthur791004 committed:
Fix typo and update changelog
